### PR TITLE
Fixes adding a new bancontact or iDeal payment method on payment method page.

### LINKED ISSUE #1929: Unable to add bancontact method with UPE from my account -> payment methods
when adding a new payment method in the my account page with UPE enabled it is not possible to use bancontact.

**To Reproduce**
1. Go to My Account -> Payment Methods -> Add New Payment Method
2. Select bancontact. Enter name and email.
3. On stripe page click Authorize

Being able to add a bancontact method.

### PR DESCRIPTION
**Pull request by superdav42**
Fixes adding a new bancontact or iDeal payment method on payment method page.
closes #1929

Turns out we are trying to hard to add a new payment method from a payment intent. We should just let the page load and ask Stripe if there are any new payment methods.

If we get the payment method from the setup intent id passed during the redirect it will be with the type bancontact or iDeal but if we query the customers account for all payment methods we can see the new payment method of the type sepa_debit that has the last 4 digits and a completely different payment method id. 

# Testing instructions
Go to payment methods. Try to add a new bancontact or ideal payment method. It should work.
Try again and force it to fail. It will fail with a relevant error message.

### CODE REVIEW

File: includes/class-wc-stripe-customer.php [hunk 1]
```diff
@@ -439,7 +439,7 @@ public function get_payment_methods( $payment_method_type ) {
 		$payment_methods = get_transient( self::PAYMENT_METHODS_TRANSIENT_KEY . $payment_method_type . $this->get_id() );
 
 		if ( false === $payment_methods ) {
-			$params   = WC_Stripe_UPE_Payment_Method_Sepa::STRIPE_ID === $payment_method_type ? '?expand[]=data.sepa_debit.generated_from.charge' : '';
+			$params   = WC_Stripe_UPE_Payment_Method_Sepa::STRIPE_ID === $payment_method_type ? '?expand[]=data.sepa_debit.generated_from.charge&expand[]=data.sepa_debit.generated_from.setup_attempt' : '';
```
Comment (FangedParakeet) [inline includes/class-wc-stripe-customer.php:442]: Nice, that `data.sepa_debit.generated_from.setup_attempt` was the missing piece of the puzzle. This is a bit pedantic, but what would you think about moving these long expandable keys--`data.sepa_debit.generated_from.charge` and `data.sepa_debit.generated_from.setup_attempt`--to a class constant? It looks like it's getting a little crowded on this line.

File: includes/class-wc-stripe-sepa-payment-token.php [hunk 2]
```diff
@@ -20,7 +20,7 @@ class WC_Payment_Token_SEPA extends WC_Payment_Token {
 	/**
 	 * Stores payment type.
 	 *
-	 * @protected string
+	 * @var string
```
Comment (FangedParakeet) [inline includes/class-wc-stripe-sepa-payment-token.php:23]: Whoops, nice fix on the annotation here.

File: includes/class-wc-stripe-sepa-payment-token.php [hunk 3]
```diff
@@ -30,16 +30,10 @@ class WC_Payment_Token_SEPA extends WC_Payment_Token {
 	 * @var array
 	 */
 	protected $extra_data = [
-		'last4' => '',
+		'last4'               => '',
+		'payment_method_type' => 'sepa_debit',
```
Comment (FangedParakeet) [inline includes/class-wc-stripe-sepa-payment-token.php:34]: Also, nice one putting this property in a more suitable place.

**Review by FangedParakeet — APPROVED** [review]
This worked great for me and was so simple, I wanted to actually kick myself for not thinking of such a simple approach. Since we can get the original payment method type from that `generated_from` field when we query Stripe for SEPA payment methods, we don't need to store anything when we actually save the payment method itself.

One small note is that to actually get any of this to work I had to set `is_reusable` on the Bancontact and iDEAL payment methods ([here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/payment-methods/class-wc-stripe-upe-payment-method-bancontact.php#L22) and [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/payment-methods/class-wc-stripe-upe-payment-method-ideal.php#L22) respectively) to `true`. Is there any reason why you skipped doing that? I'm not sure if there has been a decision made about this somewhere.

Also, on a more general note, I wonder if we could delete [this entire section](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L790-L795) from the regular checkout flow and instead just clear the cache instead, as you do [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/14b005c41503257789e36a7501d908c4ed465968/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L685-L686). If we got rid of that section entirely, I wonder if we'd even need to expand the `payment_method` in the query to retrieve the intent from Stripe. I'm not sure that we even need to save the payment method to the order at all at that point, since I don't think the subscription payments flow treads on that function (I think it goes through the parent gateway's `process_payment` function instead). Let me know your thoughts on any of that and--if you agree--whether you think it's worth making that change in this PR or in a separate issue.

Intrigued to learn your opinions on the suggestions I've made, but other than that LGTM!

**Comment by superdav42** [comment]
> One small note is that to actually get any of this to work I had to set is_reusable on the Bancontact and iDEAL payment methods (here and here respectively) to true. Is there any reason why you skipped doing that? I'm not sure if there has been a decision made about this somewhere.

Yea, I just noticed that. They were set to false when subscriptions support was added it #03556cd with comment "Disable iDeal and Bancontact for off session payments as they're both not working with the "Add payment method" page yet". So I think were safe to switch them back to true now. 

> Also, on a more general note, I wonder if we could delete this entire section from the regular checkout flow and instead just clear the cache instead, as you do here. If we got rid of that section entirely, I wonder if we'd even need to expand the payment_method in the query to retrieve the intent from Stripe. I'm not sure that we even need to save the payment method to the order at all at that point, since I don't think the subscription payments flow treads on that function (I think it goes through the parent gateway's process_payment function instead). Let me know your thoughts on any of that and--if you agree--whether you think it's worth making that change in this PR or in a separate issue.

I think we should address it separately and maybe after the release. I'm not familiar with subscriptions enough to stop storing the payment method in the order yet.
 
>  Since we can get the original payment method type from that generated_from field when we query Stripe for SEPA payment methods, we don't need to store anything when we actually save the payment method itself.

I  think  it could still be useful to have  it saved in case we ever want to change how they are displayed.

**Pull request merged**